app/src/main/java/com/rkm/tasky/network/authentication/implementation/AuthenticationManagerImpl.kt

```diff
@@ -0,0 +1,75 @@
+package com.rkm.tasky.network.authentication.implementation
+
+import com.rkm.tasky.di.IoDispatcher
+import com.rkm.tasky.network.authentication.abstraction.AuthenticationManager
+import com.rkm.tasky.network.model.dto.asAuthInfo
+import com.rkm.tasky.network.repository.implementation.AuthenticationRepositoryImpl
+import com.rkm.tasky.network.util.NetworkError
+import com.rkm.tasky.util.result.EmptyResult
+import com.rkm.tasky.util.result.asEmptyDataResult
+import com.rkm.tasky.util.result.onSuccess
+import com.rkm.tasky.util.storage.implementation.SessionStorageImpl
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class AuthenticationManagerImpl @Inject constructor(
+    private val repository: AuthenticationRepositoryImpl,
+    private val sessionStorage: SessionStorageImpl,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+) : AuthenticationManager {
+    override suspend fun logIn(
+        email: String,
+        password: String
+    ): EmptyResult<NetworkError.AuthError> =
+        withContext(dispatcher) {
+            val result = repository.loginUser(email = email, password = password)
+            result.onSuccess { user ->
+                sessionStorage.setSession(user.asAuthInfo())
+            }
+            return@withContext result.asEmptyDataResult()
+        }
+
+    override suspend fun logOut(): EmptyResult<NetworkError.AuthError> = withContext(dispatcher) {
+        val result = repository.logoutUser()
+        result.onSuccess {
+            sessionStorage.clearSession()
+        }
+        return@withContext result.asEmptyDataResult()
+    }
+
+    override suspend fun checkAuthentication(): EmptyResult<NetworkError.AuthError> =
+        withContext(dispatcher) {
+            return@withContext repository.checkAuthentication().asEmptyDataResult()
+        }
+
+    override suspend fun getNewAccessToken(): EmptyResult<NetworkError.AuthError> =
+        withContext(dispatcher) {
+            val authInfo = sessionStorage.getSession()
+            val result = repository.getNewAccessToken(
+                refreshToken = authInfo!!.refreshToken,
+                userId = authInfo.userId
+            )
```

Comment:
Would it not make sense to account for the situation where the `authInfo` here is null? You could have a situation where there is no session information in your shared preferences. In that case your app will crash with a `NullPointerException` due to that double-bang (`!!`). This can easily happen when the user clears the app’s data on their device.